gluten-core/src/main/scala/io/glutenproject/extension/PullOutProject.scala

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.extension
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.sql.shims.SparkShimLoader
+import io.glutenproject.utils.{LogicalPlanSelector, PullOutProjectHelper}
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, AggregateFunction}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern._
+import org.apache.spark.sql.execution.aggregate.TypedAggregateExpression
+
+/**
+ * This rule will insert a pre-project in the child of operators such as Aggregate, Sort, Join,
+ * etc., when they involve expressions that need to be evaluated in advance.
+ */
+case class PullOutProject(session: SparkSession)
+  extends Rule[LogicalPlan]
+  with PullOutProjectHelper {
+
+  /**
+   * Check if the input logical plan needs to add a pre-project. Different operators have different
+   * checking logic.
+   */
+  private def needsPreProject(plan: LogicalPlan): Boolean = plan match {
+    case Aggregate(groupingExpressions, aggregateExpressions, _) =>
+      groupingExpressions.exists(isNotAttribute) ||
+      aggregateExpressions.exists(_.find {
+        case ae: AggregateExpression
+            if ae.aggregateFunction.isInstanceOf[TypedAggregateExpression] =>
+          // We cannot pull out the children of TypedAggregateExpression to pre-project,
+          // and Gluten cannot support TypedAggregateExpression.
+          false
+        case ae: AggregateExpression
+            if ae.filter.exists(isNotAttribute) || ae.aggregateFunction.children.exists(
+              isNotAttributeAndLiteral) =>
+          true
+        case _ => false
+      }.isDefined)
+    case Sort(order, _, _) =>
```

Comment:
@liujiayi771 We have added a sort check in the `needPreProject` method. However, it appears that the logic for handling sort operators is not being added in this context [here](https://github.com/oap-project/gluten/pull/4245/files#diff-fa6b47ebbfa2f2ef8742dc00ce6ca270e050eb6e9d731704b0f0c5d89db37b9dR114. ).

Reply:
Could you provide more information about the `outputOrdering` issue you mentioned? Maybe i missed some context. Thanks.

Reply:
@JkSelf You can check this [discussion](https://github.com/oap-project/gluten/discussions/4308#discussioncomment-8036736).